omnipaxos/src/omni_paxos.rs

```diff
@@ -67,31 +64,23 @@ impl OmniPaxosConfig {
             !self.peers.contains(&self.pid),
             "Peers should not include self pid"
         );
-
-        let num_nodes = self.peers.len() + 1;
-        match (self.leader_quorum_size, self.append_quorum_size) {
-            (Some(prepare_quorum_size), Some(accept_quorum_size)) => {
-                let quorums_overlap = prepare_quorum_size + accept_quorum_size > num_nodes;
-                assert!(
-                    quorums_overlap,
-                    "The sum of quorum sizes must exceed the # of nodes"
-                );
-                assert!(
-                    prepare_quorum_size > 0 && prepare_quorum_size <= num_nodes,
-                    "Prepare quorum must be in range 0 to # of nodes in the cluster"
-                );
-                // TODO: currently accept_quorum_size = 1 doesn't work, add to check?
-                assert!(
-                    accept_quorum_size > 0 && accept_quorum_size <= num_nodes,
-                    "Accept quorum must be in range 0 to # of nodes in the cluster"
-                );
-                assert!(prepare_quorum_size >= accept_quorum_size, "Leader election cannot guarantee progress if prepare quorum is smaller than accept quorum");
-            }
-            (None, Some(_)) => panic!("Prepare quorum must also be defined"),
-            (Some(_), None) => panic!("Accept quorum must also be defined"),
-            (None, None) => (),
+        if let Some((prepare_quorum_size, accept_quorum_size)) = self.flexible_quorum {
+            let num_nodes = self.peers.len() + 1;
+            let quorums_overlap = prepare_quorum_size + accept_quorum_size > num_nodes;
+            assert!(
+                quorums_overlap,
+                "The quorums must overlap i.e., the sum of their sizes must exceed the # of nodes"
+            );
+            assert!(
+                prepare_quorum_size >= 1 && prepare_quorum_size <= num_nodes,
+                "Read quorum must be in range 1 to # of nodes in the cluster"
+            );
+            assert!(
+                accept_quorum_size >= 2 && accept_quorum_size <= num_nodes,
+                "Write quorum must be in range 2 to # of nodes in the cluster"
+            );
+            assert!(prepare_quorum_size >= accept_quorum_size, "Leader election cannot guarantee progress if read quorum is smaller than write quorum");
```

Comment:
This last assert is not needed. We can have a smaller prepare quorum e.g., N=5, prepare_quorum=2, accept_quorum=4 is fine.

Reply:
If we allow this then the flexible quorum fault tolerance of `accept_quorum - 1` no longer holds. In your example,  the flexible quorum should have a fault tolerance of 4-1=3 but even just 2 nodes failing means that an accept quorum cannot be formed.

But the error message doesn't make sense since the updated BLE that takes into account both quorum sizes. Maybe the check should just be what they have in the flexible quorum paper `accept_quorum_size < num_nodes / 2`? Otherwise the fault tolerance guarantee becomes `min(accept_q - 1, num_nodes - accept_q)` I think.